Generic Map Popups + spy glasses

FatCat1978 opened the pull request:
yeah u know popups? Now you can have them with screen objs.

This has a bunch of potential - both for refactoring existing systems, and new things - of which is expanded upon here: https://github.com/vgstation-coders/vgstation13/pull/24352

I've got no clue about tg code, but there's probably a way to improve this with components. it was written without them in mind.

To provide a usecase/example, simple "spy glasses" will be added. featuring a bug item, which acts as a camera, and the glasses - which, when linked, let you see the linked bug in a radius around it. A similar system could be done for AI multi-cam when the time comes. this costs nothing, and is actually detective gear.

FatCat1978 committed:
Map initial

Pacmandevil commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,86 @@
+
+			else
+				//error("[adding]'s screen_data has an invalid length. should be either 4,2,0 - it is [len]")
+				continue
```
Might be worthwhile to scrap this entirely and just use byond's shitty system, it's come in handy a few times, but getting all the legacy stuff is inevitably going to be a pain

Pacmandevil commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,86 @@
+			if(0) //legacy - screen_loc is already set up. don't add the map here, assumed to be old HUD code, or some custom overwrite (eg, x TO y) so it'd probably break it.
+
+			else
+				//error("[adding]'s screen_data has an invalid length. should be either 4,2,0 - it is [len]")
```
Also dunno what to replace this with

Qustinnus replied:
blaze it

Pacmandevil commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,86 @@
+/client/
+	var/list/screen_maps = list() //assoc list with all the active maps - when a screen obj is added to a map, it's put in here as well. "mapname" = list(screen objs in map)
+
+/obj/abstract/screen
```
wait do you guys even use these

Pacmandevil replied:
FUCK you don't

FatCat1978 committed:
removes abstract

ok

Pacmandevil commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,86 @@
+	background.assigned_map = newmap
+	background.screen_loc = "[newmap]:1,1 TO [width],[height]"
+	background.icon = 'icons/mob/actions.dmi' //change the icon to a proper one. this'll look like SHIT.
+	background.icon_state = "bg_default"
```
Note while this does compile, there's no icon.

This should be changed once I get a proper background sprite for it.

ninjanomnom replied:
You can use the same sprite you did in vg

cacogen commented:
pop windows suck

Pacmandevil commented:
and yet they're the only viable way to do UI that doesn't also block the screen.

These also use screen objs, which negate 3 completely. these will not scroll.

this type of system is also already in use in tg, and this simply serves to make it generic.

cacogen commented:
Not to be a dick or needlessly argumentative but how doesn't a pop-up window block the screen? They steal mouse focus, and there's not a lot of room to move them. If you do, they're either covering the panel in the top right or the chat in the bottom right, or resized to such a small size that they're a pain to access. And if I remember correctly, TGUI windows don't work well with multiple screens. So unless your monitor/resolution is exceptionally huge I don't see how they don't block the screen.

Pacmandevil commented:
Ok, say this system is used for something like cameras. use your brain here.

Would you rather have:

A: your view changed to the camera, taking up the entire screen

B: a popup window with the camera, which can support multiple movable windows, which can better utilize things such as multiple monitors to take up minimal actual game area.

Say you want a tetris inventory. there is 0 good way to do that without a system like this - as it'd take up screen real estate on the main map.

ninjanomnom commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,86 @@
+/client/
```
```suggestion
/client
```

ninjanomnom commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,86 @@
+/client/
+	var/list/screen_maps = list() //assoc list with all the active maps - when a screen obj is added to a map, it's put in here as well. "mapname" = list(screen objs in map)
```
This list should be created when needed

ninjanomnom commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,86 @@
+	var/list/screen_maps = list() //assoc list with all the active maps - when a screen obj is added to a map, it's put in here as well. "mapname" = list(screen objs in map)
+
+/obj/screen
+	var/assigned_map = null
```
```suggestion
	var/assigned_map
```

ninjanomnom commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,86 @@
+
+/client/proc/clear_all_popups(var/map_to_clear = null)//not really needed most of the time, as the client's screen list gets reset on relog. any of the buttons are going to get caught by garbage collection anyway. they're effectively qdel'd.
+	if(!map_to_clear)
+		return FALSE
```
Wouldn't it make more sense to just remove all of them if no arg is given by default?

Pacmandevil replied:
I've split it into two procs, one explicit "clear every popup" and one "clear x popup" so there's no situation down the line where a null somehow gets in there and fucks up everything.

FatCat1978 committed:
cleanup+Ninja fixes

FatCat1978 committed:
Update map_popups.dm

Fixes some runtimes.

AnturK commented:
I don't see much point to this, secondary map windows will not be uniform in function.
Also dmf changes are unnecessary you can just winclone new window.

Pacmandevil commented:
the function of a map window is determined entirely by it's screen objs. this just provides a standard way to create the window, add to it, and remove from it.

ninjanomnom commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,94 @@
+			if(0) //legacy - screen_loc is already set up. don't add the map here, assumed to be old HUD code, or some custom overwrite (eg, x TO y) so it'd probably break it.
+
+			else
+				//error("[adding]'s screen_data has an invalid length. should be either 4,2,0 - it is [len]")
```
You can use `CRASH()` which is a built in if you want it to runtime here and cancel the proc
You can use `stack_trace()` if you want it to error and continue the proc

FatCat1978 committed:
Update map_popups.dm

ninjanomnom commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,94 @@
+
+			else
+				CRASH("[adding]'s screen_data has an invalid length. should be either 4,2,0 - it is [len]")
+				continue
```
crash halts the proc, you wont get to this continue

Pacmandevil replied:
wait, shit. my bad.

Pacmandevil commented:
About winclone: There was something fucky about it when I initially wrote it. I had issues with the window not appearing on high (simulated) latency.

FatCat1978 committed:
coded cringe

gonna lose

FatCat1978 committed:
WIP

Adds spyglasses, fixes a pretty major bug.

FatCat1978 committed:
more bugfixes

FatCat1978 committed:
more bugfixes

Pacmandevil commented:
Alright, it's about done - some important bugs were fixed, and the glasses are about 90% done with some issues remaining.

Moving the "bug" causes the vis contents to be cut and not readded (no clue), and there's some rendering issues with lighting, shown [here: ](https://cdn.discordapp.com/attachments/484170915253321734/634805637087232061/image.png)

the white bits are lighting overlays, which are massively fucked. Curiously, they only show up in the popups. help would greatly be appreciated - they work fine out of them, shown [here](https://cdn.discordapp.com/attachments/484170915253321734/634815006717902848/image.png)

Pacmandevil commented:
Alright, the above is actually a byond bug, which I'll report as soon as I get a test case going.

The issue is that I've got absolutely no clue how lighting works, so making a simplified version FOR a test case is something I'll need help with.

FatCat1978 committed:
fixes the lighting issue.

Pacmandevil commented:
Alright, lighting bug fixed. turned out to be a shitcode issue rather than a byond issue. The planesmaster for lighting wasn't applied. renders fine now.

Remaining issues:

Dragging the bug diagonally causes vis contents to get cut completely.
Throwing/Etc doesn't update the bug.
It doesn't update when in a bag.

optimumtact commented on code/game/objects/items/devices/spyglasses.dm:
```diff
@@ -0,0 +1,91 @@
+	if("spypopup_map" in user.client.screen_maps) //alright, the popup this object uses is already IN use, so the window is open. no point in doing any other work here, so we're good. 
+		return 
+	user.client.setup_popup("spypopup",3,3,2)
+	var/list/buglist = list(linked_bug.cam_view,linked_bug.popupmaster)
```
how does this work? you're adding it to the screen loc of the popup window, but then it somehow sees the map view?

Pacmandevil replied:
`cam_view` is updated to include every turf in a range(1) in it's vis contents.

optimumtact commented on code/game/objects/items/devices/spyglasses.dm:
```diff
@@ -0,0 +1,91 @@
+	desc = "an advanced holographic projection in the shape of a pocket protector featuring technology similar to a chameleon projector. it has a built in 360 degree camera for all your nefarious needs. Simply hitting an object with it will cause it's projection to change. Microphone not included."
+	var/obj/item/clothing/glasses/regular/spy/linked_glasses
+	var/obj/screen/cam_view
+	var/obj/screen/plane_master/lighting/popupmaster 
```
what is this even used for

AnturK replied:
plane masters only apply to screen they're on

optimumtact replied:
so the lighting overlays wouldn't apply otherwise?

i don't even understand why there would be a lighting overlay if it's using a screen object

Pacmandevil replied:
@optimumtact

Plane masters only apply to the map their screen loc is set to (normally it's just the main map, which is fine - I'm pretty sure the only other map in the game thus far is the character create)

However, this becomes an issue in this specific circumstance, as when turfs are added to vis contents, **everything** on the turf is rendered along with it, *including lighting overlays*.

This results in the overlay failing to blend due to a lack of a planemaster, which looks something like [this](https://media.discordapp.net/attachments/484170915253321734/634803956241006592/image.png)

That is exactly how it'd look on the main map if the planemaster was deleted.

optimumtact replied:
yeah anturk explained in discord but thanks as well

81Denton commented:
conflict

Pacmandevil commented:
Yeah, and I need to incorporate #47241 into the spyglasses. Likely able to do this some time next week if all goes well.

FatCat1978 committed:
Merge branch 'master' into get-that-fukkin-shirt

FatCat1978 committed:
fixes tracking/reorganizes it

just need to do the polishing now. I'll do it when I wake up.

FatCat1978 committed:
fixes and shit

should be good to go now. didn't impliment the cloning, mostly because I'd feel it'd be kinda OP.

ninjanomnom commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,98 @@
+	var/del_on_map_removal = TRUE//this could probably be changed to be a proc, for conditional removal. for now, this works.
+
+/client/proc/clear_map(var/map_to_clear)//not really needed most of the time, as the client's screen list gets reset on relog. any of the buttons are going to get caught by garbage collection anyway. they're effectively qdel'd.
+	if(!map_to_clear|| !(map_to_clear in screen_maps))
```
```suggestion
	if(!screen_maps[map_to_clear])
```

Pacmandevil replied:
yeah this causes runtimes

ninjanomnom commented on code/game/objects/items/devices/spyglasses.dm:
```diff
@@ -0,0 +1,81 @@
+	var/obj/item/spy_bug/newbug = new(src)
+	var/obj/item/clothing/glasses/regular/spy/newglasses = new(src)
+	newbug.linked_glasses = newglasses
+	newglasses.linked_bug = newbug
```
Add a piece of paper explaining how to use these to the box

ninjanomnom commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,98 @@
+		screen_maps[map_to_clear] -= x
+		if(x.del_on_map_removal)
+			qdel(x)
+	screen_maps-= map_to_clear
```
```suggestion
	screen_maps -= map_to_clear
```

ninjanomnom commented on code/_onclick/hud/map_popups.dm:
```diff
@@ -0,0 +1,98 @@
+	handle_popup_close(popup)
+
+/client/verb/handle_popup_close(window_id as text) //when the popup closes in any way (player or proc call) it calls this.
+	set hidden = 1
```
```suggestion
	set hidden = TRUE
```

ninjanomnom commented on code/game/objects/items/devices/spyglasses.dm:
```diff
@@ -0,0 +1,81 @@
+	if(!user)
+		return
+	if(!user.client)
+		return	
```
You can combine these
`if(!user?.client)`

Pacmandevil replied:
I guess, I generally don't like doing that because it's less readable, but if it's requested I'll do it.

ninjanomnom commented on code/game/objects/items/devices/spyglasses.dm:
```diff
@@ -0,0 +1,81 @@
+	//if there's ever a way to make planesmasters omnipresent, then this wouldn't be needed.
+
+/obj/item/spy_bug/Destroy()
+	. = ..()
```
Destroy is an exception and parent should be called last as
`return ..()`

ninjanomnom commented on code/game/objects/items/devices/spyglasses.dm:
```diff
@@ -0,0 +1,81 @@
+/obj/item/spy_bug/Destroy()
+	. = ..()
+	qdel(tracker)
+
```
Clean up your other references here like linked_glasses

ninjanomnom commented on code/game/objects/items/devices/spyglasses.dm:
```diff
@@ -0,0 +1,81 @@
+//antag spyglasses. meant to be an example for map_popups.dm
+/obj/item/clothing/glasses/regular/spy
+	desc = "Made by Nerd. Co's infiltration and surveillance department. Upon closer inspection, there's a small screen in each lens."
+	var/obj/item/spy_bug/linked_bug
```
This needs to be cleaned up in destroy

ninjanomnom commented on code/game/objects/items/devices/spyglasses.dm:
```diff
@@ -0,0 +1,81 @@
+	cam_view.assigned_map = "spypopup_map"
+	cam_view.screen_info = list(1,1)
+
+	popupmaster = new
```
use the existing lighting plane master instead of making a new one here

ninjanomnom replied:
nevermind, ignore me. I forgot how our planemasters worked

FatCat1978 committed:
Merge branch 'master' into get-that-fukkin-shirt

FatCat1978 committed:
feedback for dummies

Fuck the entire "For Dummies" series of books seriously the way that dumb fucker on the cover looks pisses me off it's like his eyes are some shitty anime girl but without any of the ass or thighs present on normal anime girls, literally all the minuses without a smidgen of a plus god FUCK I hate him.  He knows it too, the smug bastard is smiling. "tee hee" he thinks, point up at some text we OBVIOUSLY can't see without noticing his smug ass mentioning it. "What a dumbass, can't even see the text" well guess what, idiot. I CAN see it, and I hate you all the more for it.

FatCat1978 committed:
:bangbang:

FatCat1978 committed:
Merge branch 'master' into get-that-fukkin-shirt

FatCat1978 committed:
background

remember when this was just to get a hacktoberfest tshirt lmao
